File: sections/72_master_flow_convex_mixed/problems/f_congested_team_assignment/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
const int64 INF = (1LL << 62);

struct MCF {
  struct Edge {
    int to, rev, cap, initial;
    int64 cost;
  };
  vector<vector<Edge>> g;
  MCF(int n) : g(n) {}
  pair<int, int> add_edge(int u, int v, int cap, int64 cost) {
    int id = g[u].size();
    g[u].push_back({v, (int)g[v].size(), cap, cap, cost});
    g[v].push_back({u, id, 0, 0, -cost});
    return {u, id};
  }
  pair<int, int64> flow(int s, int t, int limit) {
    int n = g.size(), sent = 0;
    int64 cost = 0;
    vector<int64> pot(n, INF), dist(n);
    pot[s] = 0;
    for (int rep = 0; rep < n; ++rep) {
      bool changed = false;
      for (int u = 0; u < n; ++u)
        if (pot[u] < INF)
          for (auto &e : g[u])
            if (e.cap && pot[e.to] > pot[u] + e.cost)
              pot[e.to] = pot[u] + e.cost, changed = true;
      if (!changed)
        break;
    }
    for (auto &x : pot)
      if (x == INF)
        x = 0;
    vector<int> pv(n), pe(n);
    while (sent < limit) {
      fill(dist.begin(), dist.end(), INF);
      dist[s] = 0;
      priority_queue<pair<int64, int>, vector<pair<int64, int>>,
                     greater<pair<int64, int>>>
          pq;
      pq.push({0, s});
      while (!pq.empty()) {
        auto [d, u] = pq.top();
        pq.pop();
        if (d != dist[u])
          continue;
        for (int i = 0; i < (int)g[u].size(); ++i) {
          auto &e = g[u][i];
          int64 nd = d + e.cost + pot[u] - pot[e.to];
          if (e.cap && nd < dist[e.to])
            dist[e.to] = nd, pv[e.to] = u, pe[e.to] = i, pq.push({nd, e.to});
        }
      }
      if (dist[t] == INF)
        break;
      for (int v = 0; v < n; ++v)
        if (dist[v] < INF)
          pot[v] += dist[v];
      int add = limit - sent;
      for (int v = t; v != s; v = pv[v])
        add = min(add, g[pv[v]][pe[v]].cap);
      for (int v = t; v != s; v = pv[v]) {
        Edge &e = g[pv[v]][pe[v]];
        cost += (int64)add * e.cost;
        e.cap -= add;
        g[v][e.rev].cap += add;
      }
      sent += add;
    }
    return {sent, cost};
  }
};

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, p, m;
  if (!(cin >> n >> p >> m))
    return 0;
  vector<int> low(p), high(p), a(p);
  int required = 0, max_a = 0;
  for (int j = 0; j < p; ++j)
    cin >> low[j] >> high[j] >> a[j], required += low[j],
        max_a = max(max_a, a[j]);
  struct Choice {
    int w, j, profit;
  };
  vector<Choice> choices(m);
  int max_profit = 0;
  for (auto &[w, j, profit] : choices)
    cin >> w >> j >> profit, --w, --j, max_profit = max(max_profit, profit);
  int s = n + p, t = s + 1;
  MCF mcf(t + 1);
  for (int w = 0; w < n; ++w)
    mcf.add_edge(s, w, 1, 0);
  for (auto [w, j, profit] : choices)
    mcf.add_edge(w, n + j, 1, -profit);
  int64 big = (int64)(n + 1) * (max_profit + 2LL * max_a * n + 1);
  vector<pair<int, int>> mandatory;
  for (int j = 0; j < p; ++j)
    for (int k = 1; k <= high[j]; ++k) {
      int64 cost = (int64)a[j] * (2LL * k - 1);
      auto ref = mcf.add_edge(n + j, t, 1, cost - (k <= low[j] ? big : 0));
      if (k <= low[j])
        mandatory.push_back(ref);
    }
  auto [sent, cost] = mcf.flow(s, t, n);
  bool ok = sent == n;
  for (auto [u, id] : mandatory)
    ok &= mcf.g[u][id].cap == 0;
  if (!ok)
    cout << "IMPOSSIBLE\n";
  else
    cout << -cost - big * required << '\n';
}

File: sections/72_master_flow_convex_mixed/problems/f_congested_team_assignment/solution.py
import heapq
import sys


class MCF:
    def __init__(self, n):
        self.g = [[] for _ in range(n)]

    def add_edge(self, u, v, cap, cost):
        ref = (u, len(self.g[u]))
        self.g[u].append([v, cap, cost, len(self.g[v]), cap])
        self.g[v].append([u, 0, -cost, len(self.g[u]) - 1, 0])
        return ref

    def flow(self, s, t, limit):
        n = len(self.g)
        inf = 10**40
        potential = [inf] * n
        potential[s] = 0
        for _ in range(n):
            changed = False
            for u in range(n):
                if potential[u] == inf:
                    continue
                for v, cap, cost, _, _ in self.g[u]:
                    if cap and potential[v] > potential[u] + cost:
                        potential[v] = potential[u] + cost
                        changed = True
            if not changed:
                break
        potential = [0 if x == inf else x for x in potential]
        sent = total = 0
        while sent < limit:
            dist = [inf] * n
            prev = [None] * n
            dist[s] = 0
            pq = [(0, s)]
            while pq:
                d, u = heapq.heappop(pq)
                if d != dist[u]:
                    continue
                for i, (v, cap, cost, _, _) in enumerate(self.g[u]):
                    nd = d + cost + potential[u] - potential[v]
                    if cap and nd < dist[v]:
                        dist[v] = nd
                        prev[v] = (u, i)
                        heapq.heappush(pq, (nd, v))
            if dist[t] == inf:
                break
            for v in range(n):
                if dist[v] < inf:
                    potential[v] += dist[v]
            add = limit - sent
            v = t
            while v != s:
                u, i = prev[v]
                add = min(add, self.g[u][i][1])
                v = u
            v = t
            while v != s:
                u, i = prev[v]
                e = self.g[u][i]
                total += add * e[2]
                e[1] -= add
                self.g[v][e[3]][1] += add
                v = u
            sent += add
        return sent, total


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, p, m = data[:3]
    at = 3
    projects = []
    for _ in range(p):
        projects.append(tuple(data[at : at + 3]))
        at += 3
    choices = []
    max_profit = 0
    for _ in range(m):
        w, j, profit = data[at : at + 3]
        at += 3
        choices.append((w - 1, j - 1, profit))
        max_profit = max(max_profit, profit)
    s, t = n + p, n + p + 1
    mcf = MCF(t + 1)
    for w in range(n):
        mcf.add_edge(s, w, 1, 0)
    for w, j, profit in choices:
        mcf.add_edge(w, n + j, 1, -profit)
    max_a = max(a for _, _, a in projects)
    big = (n + 1) * (max_profit + 2 * max_a * n + 1)
    mandatory = []
    for j, (low, high, a) in enumerate(projects):
        for k in range(1, high + 1):
            ref = mcf.add_edge(
                n + j, t, 1, a * (2 * k - 1) - (big if k <= low else 0)
            )
            if k <= low:
                mandatory.append(ref)
    sent, cost = mcf.flow(s, t, n)
    ok = sent == n and all(mcf.g[u][i][1] == 0 for u, i in mandatory)
    print(
        -cost - big * sum(low for low, _, _ in projects)
        if ok
        else "IMPOSSIBLE"
    )


if __name__ == "__main__":
    main()
